hail/python/hailtop/auth/sql_config.py

```diff
@@ -1,22 +1,94 @@
+from typing import Dict, NamedTuple, Optional, Any
 import json
+import os
 
 
-def create_secret_data_from_config(config, server_ca, client_cert, client_key):
-    assert all(config.get(x) is not None
-               for x in ('host', 'user', 'password', 'db', 'ssl-ca', 'ssl-cert',
-                         'ssl-key', 'ssl-mode')), config.keys()
-    secret_data = dict()
-    secret_data['sql-config.json'] = json.dumps(config)
-    secret_data['sql-config.cnf'] = f'''[client]
-host={config["host"]}
-user={config["user"]}
-password="{config["password"]}"
-database={config["db"]}
-ssl-ca={config["ssl-ca"]}
-ssl-cert={config["ssl-cert"]}
-ssl-key={config["ssl-key"]}
-ssl-mode={config["ssl-mode"]}
+class SQLConfig(NamedTuple):
+    host: str
+    port: int
+    user: str
+    password: str
+    instance: str
+    connection_name: str
+    db: Optional[str]
+    ssl_ca: str
+    ssl_cert: str
+    ssl_key: str
+    ssl_mode: str
+
+    def to_json(self) -> str:
+        return json.dumps(self.to_dict())
+
+    def to_dict(self) -> Dict[str, Any]:
+        d = {'host': self.host,
+             'port': self.port,
+             'user': self.user,
+             'password': self.password,
+             'instance': self.instance,
+             'connection_name': self.connection_name,
+             'ssl-ca': self.ssl_ca,
+             'ssl-cert': self.ssl_cert,
+             'ssl-key': self.ssl_key,
+             'ssl-mode': self.ssl_mode}
+        if self.db is not None:
+            d['db'] = self.db
+        return d
+
+    def to_cnf(self) -> str:
+        host_user_password = f'''[client]
+host={self.host}
+user={self.user}
+password="{self.password}"
+'''
+        database_setting = f'database={self.db}\n' if self.db is not None else ''
+        ssl_settings = f'''ssl-ca={self.ssl_ca}
+ssl-cert={self.ssl_cert}
+ssl-key={self.ssl_key}
+ssl-mode={self.ssl_mode}
 '''
+        return host_user_password + database_setting + ssl_settings
+
+    def check(self):
+        assert self.ssl_ca is not None
+        assert self.ssl_cert is not None
+        assert self.ssl_key is not None
+        if not os.path.isfile(self.ssl_cert):
+            raise ValueError(f'specified ssl-cert, {self.ssl_cert}, does not exist')
+        if not os.path.isfile(self.ssl_key):
+            raise ValueError(f'specified ssl-key, {self.ssl_key}, does not exist')
+        if not os.path.isfile(self.ssl_ca):
+            raise ValueError(f'specified ssl-ca, {self.ssl_ca}, does not exist')
+
+    @staticmethod
+    def from_json(s: str) -> 'SQLConfig':
+        return SQLConfig.from_dict(json.loads(s))
+
+    @staticmethod
+    def from_dict(d: Dict[str, Any]) -> 'SQLConfig':
+        for k in ('host', 'user', 'password', 'ssl-ca', 'ssl-cert', 'ssl-key', 'ssl-mode'):
```

Comment:
Should this check be part of `check`?

Reply:
I added these checks to `check`, but left them here too. `check` also verifies that the referenced files exist, which I don't want to require on `from_dict` because sometimes we mount the sql_config as `/database-server-config/` even though the `sql_config`'s ssl-ca says `/sql-config/...`. We only do that when we're mounting the database-server-config to use as a template for creating dev namespace database-server-configs.